icicle/tests/test_hash_api.cpp

```diff
@@ -315,54 +322,216 @@ class HashSumBackend : public HashBackend
   }
+  return true;
 }
 
+/**
+ * @brief Wrapper to the non-template version of is_valid_tree above, allowing to insert different types of arrays as
+ * inputs. Builds tree in a straight-forward single-threaded manner and compares the result with Icicle's calculation.
+ * @param tree - Merkle tree to test.
+ * @param input_size - Size of input in bytes.
+ * @param inputs - Input as a byte array.
+ * @param hashes - Vector of hashes of each layer in the tree above.
+ * @return True if the tree's calculations (icicle and test) match.
+ */
 template <typename T>
-void assert_valid_tree(
+bool is_valid_tree(
   const MerkleTree& tree,
   int nof_inputs,
   const T* inputs,
   const std::vector<Hash>& hashes,
   const MerkleTreeConfig& config)
 {
-  return assert_valid_tree(tree, nof_inputs * sizeof(T), reinterpret_cast<const std::byte*>(inputs), hashes, config);
+  return is_valid_tree(
+    tree, nof_inputs * sizeof(T), sizeof(T), reinterpret_cast<const std::byte*>(inputs), hashes, config);
+}
+
+/**
+ * @brief Function used by the HashApiTest to test the various Merkle trees defined in the tests below. Checks
+ * validity of the tree construction, and correctness/incorrectness of valid/invalid proofs generated by the tree.
+ * @param hashes - Vector of hashes of each layer in the tree above.
+ * @param config - Merkle tree config (Mostly irrelevant for cpu tests).
+ * @param output_store_min_layer - Store layer parameter for the Merkle tree builder.
+ * @param nof_leaves - Size of the T leaves array.
+ * @param leaves - Aforementioned leaves array.
+ * @param explicit_leaf_size - Optional. Size of each leaf element in case that leaves is given as a byte array. NOTE
+ *                             test will fail if this value isn't default (1) and T != std::byte
+ */
+template <typename T>
+void test_merkle_tree(
+  const std::vector<Hash>& hashes,
+  const MerkleTreeConfig& config,
+  const int output_store_min_layer,
+  int nof_leaves,
+  const T* leaves,
+  unsigned explicit_leaf_size = 1)
+{
+  ASSERT_TRUE((explicit_leaf_size == 1 || std::is_same<T, std::byte>::value))
+    << "Explicit leaf size should only be given when the given leaves array is a bytes array.";
+
+  unsigned leaf_size = explicit_leaf_size > 1 ? explicit_leaf_size : sizeof(T);
+  auto prover_tree = MerkleTree::create(hashes, leaf_size, output_store_min_layer);
+  auto verifier_tree = MerkleTree::create(hashes, leaf_size, output_store_min_layer);
+
+  // assert that incorrect size fails
+  if (config.padding_policy == PaddingPolicy::None) {
+    ASSERT_NE(prover_tree.build(leaves, nof_leaves * explicit_leaf_size - 1, config), eIcicleError::SUCCESS);
+    ASSERT_NE(prover_tree.build(leaves, nof_leaves * explicit_leaf_size + 1, config), eIcicleError::SUCCESS);
+  }
+  // build tree
+  START_TIMER(MerkleTree_build)
+  ICICLE_CHECK(prover_tree.build(leaves, nof_leaves * explicit_leaf_size, config));
+  END_TIMER(MerkleTree_build, "Merkle Tree CPU", true)
```

Comment:
is it CPU always? you didn't set the device so it can be anything here

Reply:
Nope, it's a remanent from a previous version of the test file (when it was for CPU only). I will remove it.